cpp/src/cuda_event.cpp

```diff
@@ -37,37 +36,35 @@ CudaEvent& CudaEvent::operator=(CudaEvent&& other) {
             "cannot move into an already-initialized CudaEvent",
             std::invalid_argument
         );
-        event_ = other.event_;
-        done_.store(other.done_.load());
         other.event_ = nullptr;
     }
     return *this;
 }
 
 void CudaEvent::record(rmm::cuda_stream_view stream) {
+    std::lock_guard<std::mutex> lock(done_mutex_);
+    done_ = false;
     RAPIDSMPF_CUDA_TRY(cudaEventRecord(event_, stream));
 }
 
-[[nodiscard]] bool CudaEvent::CudaEvent::is_ready() {
-    if (!done_.load(std::memory_order_acquire)) {
-        auto result = cudaEventQuery(event_);
-        done_.store(result == cudaSuccess, std::memory_order_release);
-        if (result != cudaSuccess && result != cudaErrorNotReady) {
-            RAPIDSMPF_CUDA_TRY(result);
-        }
-        return result == cudaSuccess;
+[[nodiscard]] bool CudaEvent::CudaEvent::is_ready() const {
+    std::lock_guard<std::mutex> lock(done_mutex_);
```

Comment:
I wonder if we should just stop caching the result and always defer to `cudaEventQuery`.